os/syncd/emulab-sync.1
.TH EMULAB-SYNC 1 "April 5, 2004" "Emulab" "Emulab Commands Manual"
.OS
.SH NAME
emulab-sync \- Simple distributed synchronization client.
.SH SYNOPSIS
.BI emulab-sync
[\fB-hV\fR]
[\fB-n \fIname\fR]
[\fB-d\fR]
[\fB-s \fIserver\fR]
[\fB-p \fIport-number\fR]
[\fB-i \fIcount\fR]
[\fB-a\fR]
[\fB-u\fR]
[\fB-e \fIerror-number\fR]
.SH DESCRIPTION
The
.B emulab-sync
utility is used to wait for other programs in your experiment to reach some
specific point before proceeding forward.  This utility and its server,
.B emulab-syncd\fR,
implement this synchronization through a simple counter based barrier.  The
counter is increased by
.B emulab-sync
when using the
.B -i
option and then decremented by one on each regular use.  Once the counter
reaches zero, all of the waiters will return and execution can proceed.
.P
Multiple synchronizations can take place at the same time by using the
.B -n
option to specify a separate name for each barrier.  For example, scripts on
nodeA and nodeB can be waiting on a barrier named "foo" while (other) scripts
on nodeA and nodeC can be waiting on a barrier named "bar." You may reuse an
existing barrier (including the default barrier) once it has been released (all
clients arrived and woken up).
.P
Basic error reporting is also supported using the
.B -e
option and the program's exit code.  For example, if a client encountered an
error and would be unable to continue after crossing the barrier, it can signal
its peers by using the
.B -e
option with a non-zero error code.  Then, when all of the clients have arrived,
they will wake up and exit with this non-zero error code.  If multiple peers
encounter errors, the maximum of all of the errors will be returned.
.P
Available options:
.P
.TP
\fB-h
Print out a usage message.
.TP
\fB-V
Print out a version number.
.TP
\fB-n \fIname
The name of the barrier, must be less than 64 bytes long.  (Default: barrier)
.TP
\fB-d
Turn on debugging.
.TP
\fB-s \fIserver
Specify the host where the
.B emulab-syncd
server is running.  Normally, you do not need to use this option since Emulab
will automatically use the experiment's default sync server.  The experiment's
sync server is picked at random or specified by the
.B tb-set-sync-server
NS command when creating the experiment.
.TP
\fB-p \fIport-number
Specify the port number the server is listening on.  (Default: 16534)
.TP
\fB-i \fIcount
Increase the number of waiters by
.I count
waiters.
.B NOTE\fR:
The number of waiters should not include this call, unless you are using the
.B -a
option and intend to do another sync later on.
.TP
\fB-a
Return immediately after increasing the barrier value.  Can only be used 
with the
.B -i
option.
.TP
\fB-u
Use UDP to contact the server.  (Default: TCP)
.TP
\fB-e \fIerror-number
Specify the error code for this client, must be zero (no error) or >= 32.  
The error codes for all of the clients are collected by the server and 
the maximum is returned by the clients when they exit.  Note that we 
reserve error codes [1, 32) so using large numbers is best to avoid 
interference.  (Default: 0)
.SH RETURN VALUES
.TP
>= 32
If the barrier was crossed, but atleast one of the clients reported a non-zero
error code.
.TP
10
If the server received a SIGHUP from a user, causing all of the barriers 
to be cleared and the clients released.
.TP
1
If there was an invalid command line argument.
.TP
0
If the barrier was successfully crossed without error.
.SH EXAMPLES
.PP
To synchronize one machine with another:
.PP
.RS
[gob@bluth1 ~] emulab-sync -i 1
.P
[michael@bluth2 ~] emulab-sync
.RE
.PP
To perform two different synchronizations at the same time using barriers named
"staircar" and "prison":
.PP
.RS
[marta@bluth1 ~] emulab-sync -i 1 -n staircar
.P
[michael@bluth2 ~] emulab-sync -n staircar
.P
[gob@bluth1 ~] emulab-sync -i 1 -n prison
.P
[george@bluth2 ~] emulab-sync -n prison
.RE
.PP
To report an error 
.PP
.RS
.PD 0
[buster@bluth1 ~] emulab-sync -i 1 -e 100; echo $?
.P
100
.PD
.P
.PD 0
[annyong@bluth2 ~] emulab-sync -e 0; echo $?
.P
100
.PD
.RE
.SH NOTES
If you terminate a TCP client before it returns, the server will recognize the
closed socket and adjust the counter so that it appears that the call 
never occurred.
.SH BUGS
Despite the above efforts, it is still possible for a barrier's counter to end
up with an unknown value.  Typical symptoms of this condition are clients that
refuse to wakeup despite all of them checking in.  At this point it is probably
best to use a different barrier name or you can clear all of the barriers by
sending a SIGHUP to the experiment's
.B emulab-syncd\fR.
.SH SEE ALSO
emulab-syncd(1)
.SH AUTHOR
The Emulab project at the University of Utah.
.SH NOTES
The Emulab project can be found on the web at
.IR http://www.emulab.net


doc/ChangeLog
stoller     2003/11/17 17:30:37 MST

  Modified files:
    db                   libdb.pm.in 
    sql                  database-fill.sql 
    tbsetup              assign_wrapper.in batch_daemon.in 
                         batchexp.in endexp.in eventsys_control.in 
                         mkexpdir.in node_update.in os_setup.in 
                         startexp.in swapexp.in tbend.in 
                         tbprerun.in tbreport.in tbswap.in 
                         vnode_setup.in 
    utils                delay_config.in 
    www                  beginexp.php3 dbdefs.php3.in 
                         delaycontrol.php3 editexp.php3 
                         modifyexp.php3 showexp.php3 
                         showexp_list.php3 showstuff.php3 
                         swapexp.php3 
  Log:
  Merge the two state machines (batchstate and state) into a single
  state machine (state). All of the stuff that was previously handled by
  using batchstate is now embedded into the one state machine. Of
  course, these mostly overlapped, so its not that much of a change,
  except that we also redid the machine, adding more states (for
  example, modify phases are now explicit. To get a picture of the
  actual state machine, on boss:
  
  		stategraph -o newstates EXPTSTATE
  		gv newstates.ps
  
  Things to note:
  
  * The "batchstate" slot of the experiments table is now used solely to
    provide a lock for batch daemon. A secondary change will be to
    change the slot name to something more appropriate, but it can
    happen anytime after this new stuff is installed.
  
  * I have left expt_locked for now, but another later change will be to remove
    expt_locked, and change it to active_busy or some such new state name in
    the state machine. I have removed most uses of expt_locked, except those
    that were necessary until there is a new state to replace it.
  
  * These new changes are an implementation of the new state machine,
    but I have not done anything fancy. Most of the code is the same as
    it was before.
  
  * I suspect that there are races with the batch daemon now, but they
    are going to be rare, and the end result is probably that a
    cancelation is delayed a little bit.
  
  Revision  Changes    Path
  1.138     +30 -77    testbed/db/libdb.pm.in
  1.35      +26 -31    testbed/sql/database-fill.sql
  1.138     +5 -5      testbed/tbsetup/assign_wrapper.in
  1.41      +76 -95    testbed/tbsetup/batch_daemon.in
  1.41      +19 -41    testbed/tbsetup/batchexp.in
  1.29      +101 -57   testbed/tbsetup/endexp.in
  1.12      +1 -0      testbed/tbsetup/eventsys_control.in
  2.5       +8 -7      testbed/tbsetup/mkexpdir.in
  1.13      +4 -22     testbed/tbsetup/node_update.in
  1.86      +2 -0      testbed/tbsetup/os_setup.in
  1.59      +45 -39    testbed/tbsetup/startexp.in
  1.45      +220 -182  testbed/tbsetup/swapexp.in
  1.34      +6 -23     testbed/tbsetup/tbend.in
  1.42      +21 -40    testbed/tbsetup/tbprerun.in
  1.46      +5 -11     testbed/tbsetup/tbreport.in
  1.28      +93 -226   testbed/tbsetup/tbswap.in
  1.35      +2 -1      testbed/tbsetup/vnode_setup.in
  1.12      +6 -11     testbed/utils/delay_config.in
  1.41      +6 -1      testbed/www/beginexp.php3
  1.65      +1 -28     testbed/www/dbdefs.php3.in
  1.7       +7 -7      testbed/www/delaycontrol.php3
  1.5       +4 -4      testbed/www/editexp.php3
  1.11      +7 -1      testbed/www/modifyexp.php3
  1.75      +14 -15    testbed/www/showexp.php3
  1.80      +2 -2      testbed/www/showexp_list.php3
  1.121     +51 -63    testbed/www/showstuff.php3
  1.21      +3 -3      testbed/www/swapexp.php3

stoller     2003/11/17 15:35:33 MST

  Modified files:
    www                  tbauth.php3 toggle.php 
  Log:
  Add web login attack detection/prevention. Two changes:
  
  * Add slots to users table to track number of failures in the last N
    seconds. If a threshold is passed (currently 4 failures in the last
    minute), the web login is disabled. Note that I do not disable the
    ops shell login at this time. Aging is passive; the values are cleared
    when login is successful, or when more then one minute has passed
    since the last failure. In other words, a burst of failures will
    disable the login, but failures over time are okay.
  
  * Add login_failures table to do exactly the same as above, except it
    is on an IP basis (REMOTE_ADDR in the server). Currently the
    threshold is 8 failures in the last two minutes, at which time all
    logins from that IP are disabled.
  
  In both cases email is sent to tbops (and the user).
  
  The constants are defined at the top of www/tbauth.in, rather then as
  site variables, to avoid pounding the DB when an attack is being
  launched.
  
  To clear a user freeze, go to the user profile page and use the
  "toggle" near the bottom.
  
  To clear an IP freeze: delete from login_failures were IP='1.1.1.1'
  
  Revision  Changes    Path
  1.40      +144 -17   testbed/www/tbauth.php3
  1.13      +18 -3     testbed/www/toggle.php

	
stoller     2003/11/05 10:53:00 MST

  Modified files:
    db                   libdb.pm.in xmlconvert.in 
    tbsetup              batchexp.in eventsys_control.in 
    tbsetup/ns2ir        program.tcl sim.tcl.in 
  Log:
  Frontend and parser portion of two event system changes:
  
  * Generate a shared secret key for the event system. This key is
    stored into the DB, and passed to the node via tmcd. It is also
    stashed into a file in the experiment directory (can be accessed
    only by the project/group members). The key is used to attach a
    HMAC (hashed message authentication) to each event, which is checked
    by the receivers to ensure that the event is not bogus. More details
    on this later when I commit the event library/client changes.
  
  * Added "virt_programs" table to store info about each program object
    defined by the user. The intent is to no longer send the command

os/syncd/emulab-sync.c
/*
 * EMULAB-COPYRIGHT
 * Copyright (c) 2000-2004 University of Utah and the Flux Group.
 * All rights reserved.
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <sys/un.h>
#include <sys/fcntl.h>
#include <stdio.h>
#include <errno.h>
#include <syslog.h>
#include <unistd.h>
#include <signal.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <time.h>
#include <assert.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <setjmp.h>
#include "decls.h"
#include "config.h"

#ifndef SYNCSERVER
#define SYNCSERVER	"/var/emulab/boot/syncserver"
#endif

static int	debug = 0;

/* Forward decls */
static int	doudp(barrier_req_t *, struct in_addr, int);
static int	dotcp(barrier_req_t *, struct in_addr, int);

/* For connect timeouts. */
static jmp_buf	deadline;
static int	deadsyncdflag;
static void	deadsyncd(int sig);

char *usagestr = 
 "usage: emulab-sync [options]\n"
 " -h              Display this message\n"
 " -V              Print version information and exit\n"
 " -n name         Optional barrier name; must be less than 64 bytes long\n"
 " -d              Turn on debugging\n"
 " -s server       Specify a sync server to connect to\n"
 " -p portnum      Specify a port number to connect to\n"
 " -i count        Increase the barrier value by 'count' waiters\n"
 " -a              Master returns immediately when initializing barrier\n"
 " -u              Use UDP instead of TCP\n"
 " -e errornum     Signal an error (must be zero or >= 32)\n"
 " -m              Test whether or not this is the 'master' node\n"
 "\n";

void
usage()
{
	fprintf(stderr, "%s", usagestr);
	exit(64);
}

int
main(int argc, char **argv)
{
	int			n, ch, errval = 0;
	struct hostent		*he;
	FILE			*fp;
	struct in_addr		serverip;
	int			portnum = SERVER_PORTNUM;
	int			useudp  = 0;
	char			*server = NULL;
	int			initme  = 0;
	int			nowait  = 0;
	char			*barrier= DEFAULT_BARRIER;
	barrier_req_t		barrier_req;
	extern char		build_info[];
	int			mastercheck = 0;
	
	while ((ch = getopt(argc, argv, "hVds:p:ui:n:ae:m")) != -1)
		switch(ch) {
		case 'a':
			nowait++;
			break;
		case 'd':
			debug++;
			break;
		case 'p':
			if (sscanf(optarg, "%d", &portnum) == 0) {
				fprintf(stderr,
					"Error: -p value is not a number: "
					"%s\n",
					optarg);
				usage();
			}
			else if ((portnum <= 0) || (portnum >= 65536)) {
				fprintf(stderr,
					"Error: -p value is not between "
					"0 and 65536: %d\n",
					portnum);
				usage();
			}
			break;
		case 'i':
			if (sscanf(optarg, "%d", &initme) == 0) {
				fprintf(stderr,
					"Error: -i value is not a number: "
					"%s\n",
					optarg);
				usage();
			}
			else if (initme <= 0) {
				fprintf(stderr,
					"Error: -i value must be greater "
					"than zero: %s\n",
					optarg);
				usage();
			}
			break;
		case 'n':
			if (strlen(optarg) == 0) {
				fprintf(stderr, "Error: -n value is empty\n");
				usage();
			}
			else if (strlen(optarg) >= 64) {
				fprintf(stderr,
					"Error: -n value is too long: %s\n",
					optarg);
				usage();
			}
			else {
				barrier = optarg;
			}
			break;
		case 's':
			if (strlen(optarg) == 0) {
				fprintf(stderr, "Error: -s value is empty\n");
				usage();
			}
			else {
				server = optarg;
			}
			break;
		case 'u':
			useudp = 1;
			break;
		case 'e':
			if (sscanf(optarg, "%d", &errval) == 0) {
				fprintf(stderr,
					"Error: -e value is not a number: "
					"%s\n",
					optarg);
				usage();
			}
			else if ((errval < 0) ||
				 (errval >= SERVER_ERROR_BASE)) {
				fprintf(stderr,
					"Error: -e value must be zero or "
					"less than %d\n",
					SERVER_ERROR_BASE);
				usage();
			}
			break;
		case 'V':
			fprintf(stderr, "%s\n", build_info);
			exit(0);
			break;
		case 'h':
			fprintf(stderr, "%s", usagestr);
			exit(0);
			break;
		case 'm':
			mastercheck = 1;
			break;
		default:
			usage();
		}

	if (nowait && (initme == 0)) {
		fprintf(stderr,	"Error: -a requires the -i option\n");
		usage();
	}
	
	argv += optind;
	argc -= optind;

	if (argc) {
		fprintf(stderr,
			"Error: Unrecognized command line arguments: %s ...\n",
			argv[0]);
		usage();
	}

	/*
	 * Look for the syncserver access file.
	 */
	if (!server &&
	    (access(SYNCSERVER, R_OK) == 0) &&
	    ((fp = fopen(SYNCSERVER, "r")) != NULL)) {
		char	buf[BUFSIZ], *bp;
		
		if (fgets(buf, sizeof(buf), fp)) {
			if ((bp = strchr(buf, '\n')))
				*bp = (char) NULL;

			/*
			 * Look for port spec
			 */
			if ((bp = strchr(buf, ':'))) {
				*bp++   = (char) NULL;
				portnum = atoi(bp);
			}
			server = strdup(buf);
		}
		fclose(fp);
	}
	if (!server) {
		fprintf(stderr,
			"Error: Could not deduce the name of the syncserver!\n"
			"You may be running in an older experiment, you\n"
			"will have to manually start the emulab-syncd and\n"
			"use the -s option, or, you can recreate your\n"
			"experiment.\n\n");
		usage();
	}
	if (mastercheck) {
		char nodename[BUFSIZ];

		gethostname(nodename, sizeof(nodename));
		if (strcmp(server, nodename) == 0)
			exit(0);
		else
			exit(1);
	}
	
	/*
	 * Map server to IP.
	 */
	he = gethostbyname(server);
	if (he)
		memcpy((char *)&serverip, he->h_addr, he->h_length);
	else {
		fprintf(stderr, "gethostbyname(%s) failed\n", server); 
		exit(68);
	}

	/*
	 * Build up the request structure to send to the server.
	 */
	bzero(&barrier_req, sizeof(barrier_req));
	strcpy(barrier_req.name, barrier);
	if (initme) {
		barrier_req.request = BARRIER_INIT;
		barrier_req.count   = initme;
		if (nowait)
			barrier_req.flags = BARRIER_INIT_NOWAIT;
	}
	else
		barrier_req.request = BARRIER_WAIT;
	barrier_req.error = errval;
	
	if (useudp)
		n = doudp(&barrier_req, serverip, portnum);
	else
		n = dotcp(&barrier_req, serverip, portnum);
	
	if (n < 0)
		exit(76); /* Protocol failure */
	else
		exit(n); /* User defined error. */

	assert(0); /* Not reached. */
}

/*
 * TCP version, which uses ssl if compiled in.
 */
static int
dotcp(barrier_req_t *barrier_reqp, struct in_addr serverip, int portnum)
{
	int			sock, n, cc;
	struct sockaddr_in	name;
	char			*bp, buf[BUFSIZ];
	
	while (1) {
		/* Create socket from which to read. */
		sock = socket(AF_INET, SOCK_STREAM, 0);
		if (sock < 0) {
			perror("creating stream socket:");
			return -1;
		}

		/* Create name. */
		name.sin_family = AF_INET;
		name.sin_addr   = serverip;
		name.sin_port   = htons(portnum);

		/* For alarm. */
		deadsyncdflag = 0;
		signal(SIGALRM, deadsyncd);
		if (setjmp(deadline)) {
			alarm(0);
			signal(SIGALRM, SIG_DFL);
			goto refused;
		}
		alarm(5);

		cc = connect(sock, (struct sockaddr *) &name, sizeof(name));
		/*
		 * If we get the alarm now, no problem. We will close the
		 * connection and try again after a short wait. The server
		 * will recongnize this for what is, not recording any state
		 * until we actually send it the request below (write to sock).
		 */
		alarm(0);
		signal(SIGALRM, SIG_DFL);

		/* Success */
		if (cc == 0)
			break;

		if (errno != ECONNREFUSED && errno != ETIMEDOUT) {
			perror("connecting stream socket");
			close(sock);
			return -1;
		}
	refused:
		close(sock);
		if (debug) 
			fprintf(stderr, "Connection refused. Waiting ...\n");
		sleep(5);
	}

	n = 1;
	if (setsockopt(sock, SOL_SOCKET, SO_KEEPALIVE, &n, sizeof(n)) < 0) {
		perror("setsockopt SO_KEEPALIVE");
		goto bad;
	}

	/*
	 * Write the command to the socket and wait for the response.
	 */
	bp = (char *) barrier_reqp;
	n  = sizeof(*barrier_reqp);
	while (n) {
		if ((cc = write(sock, bp, n)) <= 0) {
			if (cc < 0) {
				perror("Writing to socket:");
				goto bad;
			}
			fprintf(stderr, "write aborted");
			goto bad;
		}
		bp += cc;
		n  -= cc;
	}

	bp = buf;
	n = 0;
	memset(buf, 0, sizeof(buf));
	while (1) {
		if ((cc = read(sock, bp, sizeof(buf) - n)) <= 0) {
			if (cc < 0) {
				perror("Reading from socket:");
				goto bad;
			}
			break;
		}
		else {
			bp += cc;
			n += cc;
		}
	}
	close(sock);

	if (n == 4) {
		return *((int *)buf);
	}
	else {
		return -1;
	}
 bad:
	close(sock);
	return -1;
}

/*
 * Not very robust ...
 */
static int
doudp(barrier_req_t *barrier_reqp, struct in_addr serverip, int portnum)
{
	int			sock, length, n, cc;
	struct sockaddr_in	name, client;
	char			buf[BUFSIZ];

	/* Create socket from which to read. */
	sock = socket(AF_INET, SOCK_DGRAM, 0);
	if (sock < 0) {
		perror("creating dgram socket:");
		return -1;
	}

	/* Create name. */
	name.sin_family = AF_INET;
	name.sin_addr   = serverip;
	name.sin_port   = htons(portnum);

	/*
	 * Write the command to the socket and wait for the response
	 */
	n  = sizeof(*barrier_reqp);
	cc = sendto(sock, barrier_reqp,
		    n, 0, (struct sockaddr *)&name, sizeof(name));
	if (cc != n) {
		if (cc < 0) {
			perror("Writing to socket:");
			return -1;
		}
		fprintf(stderr, "short write (%d != %d)\n", cc, n);
		return -1;
	}

	cc = recvfrom(sock, buf, sizeof(buf) - 1, 0,
		      (struct sockaddr *)&client, &length);

	if (cc < 0) {
		perror("Reading from socket:");
		return -1;
	}
	close(sock);

	if (cc == 4) {
		return *((int *)buf);
	}
	else {
		return -1;
	}
}

void
deadsyncd(int sig)
{
	deadsyncdflag = 1;
	longjmp(deadline, 1);
}
